# Inspect the real structural proposal and admission

Chapter 4 introduces structural discovery, in which a proposer revises the search-space specification and admitted revisions return to numerical search. The historical companion followed a proposal to allow two-cell unions. In this fresh run the single live call proposed a different revision, a derived feature. This notebook follows it through admission to the paired confirmation comparison. Read the fresh live response, then its separate admission decision. The proposed product changes the auxiliary representation; it does not retrain the predictor. Admission does not establish that the proposal improves discovery.

**Evidence.** Run 20260923T220909Z repeated the regional study end to end on two machines. `spark-fe12` ran numerical discovery with three new optimizer seeds (74101 to 74103), one live structural proposal and a fresh 200,000-observation confirmation sample. `spark-ef84` ran 70 live answer tests and nine report drafts with `claude-sonnet-5`. The population and frozen logistic predictor are those of the chapter companions; the seeds, proposal and confirmation sample are new, so these numbers should not be combined with the historical run's.

## Setup and source checks

The setup cell locates the data root, the directory whose relative layout the evidence records use. It uses `MODEL_VALIDATION_ROOT` when that variable is set. Otherwise it searches for a repository checkout containing the current directory and then for the copy installed with the `aiv` package. `R` is this run's folder under the data root, and `read` loads a JSON file from it.

`reference.json` was exported from the run's regional results and records their SHA-256 digest. The final assertion recomputes that digest, so every value used below comes from the unchanged results of this run. This notebook reads recorded outputs; rerunning it makes no language-model call.

In [1]:
from pathlib import Path
import json, hashlib, os, platform
import numpy as np
import pandas as pd

# Locate the data root: the directory whose relative layout the evidence records use.
MARKER = "book/notebooks/regional_case/companion.py"

def find_data_root():
    configured = os.environ.get("MODEL_VALIDATION_ROOT")
    if configured:
        root = Path(configured).expanduser().resolve()
        if not (root / MARKER).is_file():
            raise FileNotFoundError(f"MODEL_VALIDATION_ROOT must contain {MARKER}")
        return root, "MODEL_VALIDATION_ROOT"
    for path in (Path.cwd(), *Path.cwd().parents):
        if (path / MARKER).is_file():
            return path, "repository checkout"
    try:
        from aiv.companions import locate
    except ImportError:
        raise FileNotFoundError(
            "Companion data not found. Install the aiv package, open the notebook "
            "inside the repository checkout, or set MODEL_VALIDATION_ROOT.") from None
    return locate()

ROOT, DATA_SOURCE = find_data_root()
LIVE = ROOT / "book/experiments/regional_live"
R = ROOT / "book/experiments/regional_live/runs/20260923T220909Z"
print("Python:", platform.python_version())
print("Data root found from:", DATA_SOURCE)
print("Run:", R.relative_to(ROOT))
read=lambda p: json.loads((R/p).read_text())
results=read('regional/results.json')
reference=read('reference.json')
seeds=reference['seeds']
assert hashlib.sha256((R/'regional/results.json').read_bytes()).hexdigest()==reference['source_sha256']


Python: 3.12.3
Data root found from: repository checkout
Run: book/experiments/regional_live/runs/20260923T220909Z


## The live proposal, its admission and the comparison

The cell checks that the live call completed, then displays the saved proposal, the separate admission record and the six paired comparisons. The proposal states its change, rationale, evidence identifiers and a falsifiable prediction. Admission records whether the revision can be executed and scored under the existing contract, together with the checked feature metadata and the review's limits. The admission assertion ties the record to the proposal by its digest.

In [2]:
call=read('regional/reflection/live_call/response.json')
assert call['status']=='ok'
proposal=read('regional/reflection/response.json')
admission=read('regional/reflection/admission.json')
assert admission['admitted']
assert admission['response_sha256']==hashlib.sha256((R/'regional/reflection/response.json').read_bytes()).hexdigest()
display(proposal)
display(admission)
display(pd.DataFrame(results['comparisons']))

{'parent_specification': 'leaf-output-medoid-single-cell-v1',
 'change_type': 'derived_feature',
 'operation': 'product',
 'operands': ['business_receipt_share', 'cash_intensity_z'],
 'feature_name': 'business_receipt_cash_intensity_product',
 'rationale': "business_receipt_share (fraction of receipts declared as business activity, pre_period) and cash_intensity_z (standardized deviation of cash receipts from the account's own baseline, period_end) form the single largest pairwise fANOVA variance among all reported pairs (3.95e-05, evidence:fanova_pairs), more than double the next pair (cash_intensity_z x counterparty_concentration_z, 1.79e-05). business_receipt_share's own main effect is essentially flat (~1e-4 in magnitude, evidence:feature:business_receipt_share) while cash_intensity_z has the strongest univariate tail effect of any feature (bin_mean_brier rising to 0.042 in its top decile, main_effect 0.031, evidence:feature:cash_intensity_z). A flat marginal effect paired with a d

{'admitted': True,
 'response_sha256': '1b9c4f943eb9f5f47a8e94dac5959a082039dd2afbdb64b7c69b64f265b5a8d0',
 'version': 'leaf-output-medoid-derived-product-v1',
 'parent': 'leaf-output-medoid-single-cell-v1',
 'reviewer': 'Codex source and executability review',
 'reason': 'Product of two existing dimensionless pre-decision fields; finite on discovery data. Auxiliary may refit; base predictor stays fixed and region remains a single cell.',
 'checks': {'feature_metadata': [{'feature': 'business_receipt_share',
    'meaning': 'Declared expected share of receipts from business activity',
    'available': 'pre_period',
    'unit': 'fraction',
    'eligibility': 'Available before the monitoring decision; not a post-investigation field'},
   {'feature': 'cash_intensity_z',
    'meaning': "Standardized deviation of cash receipts from the account's own baseline",
    'available': 'period_end',
    'unit': 'standardized score',
    'eligibility': 'Available before the monitoring decision; not a 

,seed,a,b,contrast_difference,standard_error,simultaneous_ci
0,74101,reflection,continuation,0.012834,0.001590,"[0.008051940751484483, 0.017616414017858083]"
1,74101,twinning,random,0.000000,0.000000,"[0.0, 0.0]"
2,74102,reflection,continuation,-0.001141,0.001139,"[-0.0045669455524634285, 0.0022844287437013173]"
3,74102,twinning,random,0.000000,0.000000,"[0.0, 0.0]"
4,74103,reflection,continuation,-0.000902,0.001287,"[-0.004772999696062933, 0.002968597897497625]"
5,74103,twinning,random,0.006413,0.001018,"[0.003349358407588642, 0.009475817055983117]"


**Read the output.** The response proposed `business_receipt_share * cash_intensity_z`, a product of two existing dimensionless fields, citing the largest pairwise functional ANOVA variance in the packet. Admission created version `leaf-output-medoid-derived-product-v1`. It checked that both inputs are available before the monitoring decision and that the product is finite on discovery data. The auxiliary error model could be refit with the new column, while the predictor stayed fixed and regions remained single cells. The admission also records that the proposal described quintile bins as deciles. In the comparison table, reflection minus continuation is 0.012834 at seed 74101, with an interval [0.008052, 0.017616] that excludes zero. At the other seeds the differences are −0.001141 and −0.000902, and both intervals contain zero. Where twinning and random froze the same region, their difference and its standard error are exactly zero. One of three seeds therefore shows a resolved advantage for the revision; the seeds share one discovery sample, proposal and confirmation sample, so this is not a replicated finding.